src/ply.rs

```diff
@@ -56,165 +118,186 @@ struct Property {
+        return Err(InvalidInput("No format specified".into()).into());
+    }
+
+    Ok((Header {
+            elements: elements,
+            format: format.unwrap(),
+        },
+        header_len))
+}
+
+#[derive(Debug)]
+struct PointFormat {
+    position_data_type: DataType,
+    // None, if the file does not contain color.
+    color_data_type: Option<DataType>,
+}
 
 /// Opens a PLY file and checks that it is the correct format we support. Seeks in the file to the
 /// beginning of the binary data which must be (x, y, z, r, g, b) tuples.
-// TODO(hrapp): support more of PLY and maybe pull this out into a separate crate.
-pub fn open(ply_file: &Path) -> (File, i64) {
-    // hopefully, the header is shorter.
-    let mut header_bytes = vec![0u8; 4096];
-    let mut file = File::open(ply_file).unwrap();
-    file.read_exact(&mut header_bytes).unwrap();
-    let header = {
-        let (remaining_data, header) = match header(&header_bytes) {
-            nom::IResult::Done(r, h) => (r, h),
-            nom::IResult::Error(err) => panic!("Parsing error: {}", err),
-            nom::IResult::Incomplete(_) => panic!("Parsing error: Unexpected end of data."),
-        };
+fn open(ply_file: &Path) -> Result<(BufReader<File>, i64, PointFormat)> {
+    let mut file = File::open(ply_file).chain_err(|| "Could not open input file.")?;
+    let mut reader = BufReader::new(file);
+    let (header, header_len) = parse_header(&mut reader)?;
+    file = reader.into_inner();
+    file.seek(SeekFrom::Start(header_len as u64))?;
 
-        if header.format != Format::BinaryLittleEndianV1 {
-            panic!("Unsupported PLY format: {:?}", header.format);
-        }
+    if !header.has_element("vertex") {
+        panic!("Header does not have element 'vertex'");
+    }
+
+    if header.format != Format::BinaryLittleEndianV1 {
+        panic!("Unsupported PLY format: {:?}", header.format);
+    }
 
-        if header.vertex.name != "vertex" && header.vertex.properties.len() != 6 &&
-           header.vertex.properties[0].name != "x" &&
-           header.vertex.properties[1].name != "y" &&
-           header.vertex.properties[2].name != "z" &&
-           header.vertex.properties[3].name != "red" &&
-           header.vertex.properties[4].name != "green" &&
-           header.vertex.properties[4].name != "blue" {
-            panic!("PLY must contain (x,y,z,red,green,blue) tuples.");
+    let vertex = &header["vertex"];
+    if !vertex.has_property("x") || !vertex.has_property("y") || !vertex.has_property("z") {
+        panic!("PLY must contain properties 'x', 'y', 'z' for 'vertex'.");
+    }
+
+    let mut num_bytes_per_point = 0;
+    let position_data_type = header["vertex"]["x"].data_type;
+    num_bytes_per_point += position_data_type.size_in_bytes() * 3;
+
+    let color_data_type = {
+        if vertex.has_property("red") && vertex.has_property("green") &&
+           vertex.has_property("blue") {
+            Some(vertex["red"].data_type)
```

Comment:
Add a TODO that we might want to verify that the same type is used for red/green/blue, r/g/b, or x/y/z?